compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -640,7 +640,13 @@ impl<'a> Parser<'a> {
         let mut bounds = Vec::new();
         let mut negative_bounds = Vec::new();
 
-        while self.can_begin_bound() || self.token.is_keyword(kw::Dyn) {
+        while self.can_begin_bound()
+            // Continue even if we find a keyword.
+            // This is necessary for error recover on, for example, `impl fn()`.
+            //
+            // The only keyword that can go after generic bounds is `where`, so stop if it's it.
+            || (self.token.is_reserved_ident() && !self.token.is_keyword(kw::Where))
```

Comment:
Can we specifically check for the fn keyword? Or do we want it to be general on purpose?

Reply:
I don't know tbh. `fn` is the main motivation here, but having weird errors for other keywords doesn't seem right either.

Reply:
This is fine as is then.